put evm error msg into the log(#1413)

koseoyoung opened the pull request:
I changed the receipt log to include evm error message. 
But if it is better to add a new field (errorMsg), I should change a lot of stuff 
Which would be better?

dustinxie commented:
i would say adding a new field is be better
test to make sure the old log can serialize/deserialize correctly using new format

zjshen14 commented:
Per offline discussion, we want to explore how to deal with error classification via error code

zjshen14 commented on action/receipt.go:
```diff
@@ -19,6 +21,24 @@ const (
 	FailureReceiptStatus = uint64(0)
```
Let's move error code definition to https://github.com/iotexproject/iotex-proto/tree/master/proto/types

zjshen14 commented on action/protocol/execution/evm/evm.go:
```diff
@@ -250,7 +252,37 @@ func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64) (
 	remainingGas += refund
 
-	return ret, evmParams.gas, remainingGas, contractRawAddress, evmErr != nil, nil
+	if isPostBering {
```
This is so verbose. How about following?

```suggestion
errStatusCode = Unknown
switch evmErr {
	   case vm.ErrOutOfGas:
	        errStatusCode = FailureErrOutOfGas
	    ...
}
return ret, evmParams.gas, remainingGas, contractRawAddress, errStatusCode, nil 
```

zjshen14 replied:
And you may consider write a errToErrStatusCode function for this

zjshen14 commented on action/protocol/execution/evm/evm.go:
```diff
@@ -199,21 +197,22 @@ func getChainConfig() *params.ChainConfig {
 	if err := securityDeposit(evmParams, stateDB, gasLimit); err != nil {
 		log.L().Warn("unexpected error: not enough security deposit", zap.Error(err))
-		return nil, 0, 0, action.EmptyAddress, true, err
+		return nil, 0, 0, action.EmptyAddress, action.FailureReceiptStatus, err
```
Can we be more specific what the error it is for this case?

zjshen14 commented on action/protocol/execution/evm/evm.go:
```diff
@@ -199,21 +197,22 @@ func getChainConfig() *params.ChainConfig {
 	intriGas, err := intrinsicGas(evmParams.data)
 	if err != nil {
-		return nil, evmParams.gas, remainingGas, action.EmptyAddress, true, err
+		return nil, evmParams.gas, remainingGas, action.EmptyAddress, action.FailureReceiptStatus, err
```
ditto

zjshen14 commented on action/protocol/execution/evm/evm.go:
```diff
@@ -199,21 +197,22 @@ func getChainConfig() *params.ChainConfig {
 	}
 	if remainingGas < intriGas {
-		return nil, evmParams.gas, remainingGas, action.EmptyAddress, true, action.ErrOutOfGas
+		return nil, evmParams.gas, remainingGas, action.EmptyAddress, action.FailureReceiptStatus, action.ErrOutOfGas
```
ditto

zjshen14 commented on action/protocol/execution/evm/evm.go:
```diff
@@ -161,11 +161,9 @@ func ExecuteContract(
-		receipt.Status = action.SuccessReceiptStatus
-	}
+
+	receipt.Status = statusCode
```
Should this be gated by fork height?

dustinxie commented on action/protocol/execution/evm/evm.go:
```diff
@@ -199,21 +197,22 @@ func getChainConfig() *params.ChainConfig {
 
-func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64) ([]byte, uint64, uint64, string, bool, error) {
+//Error in executeInEVM is a consensus issue
+func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64, isPostBering bool) ([]byte, uint64, uint64, string, uint64, error) {
```
do not pass isPostBering as input, this won't scale if we have more height change in the future.
You can use stateDB.hu.IsPostBering() inside executeInEVM(), see comment below

dustinxie commented on action/protocol/execution/evm/evm.go:
```diff
@@ -250,7 +252,37 @@ func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64) (
 	remainingGas += refund
 
-	return ret, evmParams.gas, remainingGas, contractRawAddress, evmErr != nil, nil
+	if isPostBering {
```
if stateDB.hu.IsPostBering()

dustinxie commented on action/protocol/execution/evm/evm.go:
```diff
@@ -236,9 +235,12 @@ func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64) (
+		// The only possible consensus-error would be if there wasn't
+		// sufficient balance to make the transfer happen.
+		// Should be a hard fork (Bering)
+		if evmErr == vm.ErrInsufficientBalance && isPostBering {
```
also why we add Bering here?
this could also happen before Bering i think

koseoyoung commented:
1. changed code to use iotex-proto receipt status 
2. added initGenesis field in smartcontracttest in protocol_test.go to test the receipt status change after height update(ford fork)
3. added 2 test files (infiniteloop, cashier for bering)

koseoyoung commented:
In geth code, there are some evm error variables which are not accessible from outside. (revertedError, and so on). So I changed it to compare this with error message from geth. So now it is successfully showing the revert errorcode(106).

dustinxie commented on action/protocol/execution/evm/evm.go:
```diff
@@ -199,21 +198,23 @@ func getChainConfig() *params.ChainConfig {
-func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64) ([]byte, uint64, uint64, string, bool, error) {
+//Error in executeInEVM is a consensus issue
+func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64, raCtx protocol.RunActionsCtx, hu config.HeightUpgrade) ([]byte, uint64, uint64, string, uint64, error) {
+	isBering := hu.IsPost(config.Bering, raCtx.BlockHeight)
```
no need to pass in hu, stateDB has hu in it
isBering = stateDB.hu.IsPost()

dustinxie commented on action/protocol/execution/protocol_test.go:
```diff
@@ -176,9 +182,10 @@ func (cfg *ExecutionConfig) ExpectedReturnValue() []byte {
-	InitBalances []ExpectedBalance `json:"initBalances"`
-	Deployments  []ExecutionConfig `json:"deployments"`
-	Executions   []ExecutionConfig `json:"executions"`
+	InitGenesis  GenesisBlockHeight `json:"initGenesis"`
```
directly use IsBering bool `json:"isBering"` here?
so no need to define the GenesisBlockHeight struct above

dustinxie commented on action/protocol/execution/evm/evm.go:
```diff
@@ -199,21 +198,23 @@ func getChainConfig() *params.ChainConfig {
-func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64) ([]byte, uint64, uint64, string, bool, error) {
+//Error in executeInEVM is a consensus issue
+func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64, raCtx protocol.RunActionsCtx) ([]byte, uint64, uint64, string, uint64, error) {
+	isBering := stateDB.hu.IsPost(config.Bering, raCtx.BlockHeight)
```
only pass in raCtx.BlockHeight, instead of entire raCtx

zjshen14 commented on action/protocol/execution/evm/evm.go:
```diff
@@ -199,21 +198,23 @@ func getChainConfig() *params.ChainConfig {
 	if err := securityDeposit(evmParams, stateDB, gasLimit); err != nil {
 		log.L().Warn("unexpected error: not enough security deposit", zap.Error(err))
-		return nil, 0, 0, action.EmptyAddress, true, err
+		return nil, 0, 0, action.EmptyAddress, uint64(iotextypes.ReceiptStatus_Failure), err
```
Shall we define NotEnoughDeposit

zjshen14 commented on action/protocol/execution/evm/evm.go:
```diff
@@ -199,21 +198,23 @@ func getChainConfig() *params.ChainConfig {
 	intriGas, err := intrinsicGas(evmParams.data)
 	if err != nil {
-		return nil, evmParams.gas, remainingGas, action.EmptyAddress, true, err
+		return nil, evmParams.gas, remainingGas, action.EmptyAddress, uint64(iotextypes.ReceiptStatus_Failure), err
```
Is it actually ReceiptStatus_ErrOutOfGas?

zjshen14 commented on action/protocol/execution/evm/evm.go:
```diff
@@ -199,21 +198,23 @@ func getChainConfig() *params.ChainConfig {
 	}
 	if remainingGas < intriGas {
-		return nil, evmParams.gas, remainingGas, action.EmptyAddress, true, action.ErrOutOfGas
+		return nil, evmParams.gas, remainingGas, action.EmptyAddress, uint64(iotextypes.ReceiptStatus_Failure), action.ErrOutOfGas
```
Is it actually ReceiptStatus_ErrOutOfGas

zjshen14 commented on action/protocol/execution/evm/evm.go:
```diff
@@ -236,9 +237,12 @@ func executeInEVM(evmParams *Params, stateDB *StateDBAdapter, gasLimit uint64) (
+		// The only possible consensus-error would be if there wasn't
+		// sufficient balance to make the transfer happen.
+		// Should be a hard fork (Bering)
+		if evmErr == vm.ErrInsufficientBalance && isBering {
```
This logic seems to be changed wrongly? Before the change, if `evmErr == vm.ErrInsufficientBalance`, we will return error. After the change, if before isBering height, there will return error at all?
